CppCoreGuidelines.md

```diff
@@ -15082,6 +15084,84 @@ Use a function for coroutines.
 Flag a lambda that is a coroutine and has a non-empty capture list.
 
 
+### <a name="Rcoro-locks"></a>CP.52: Do not hold locks or other synchronization primitives across suspension points
+
+##### Reason
+
+This pattern creates a significant risk of deadlocks.  Some types of waits will allow the current thread to perform additional work until the asynchronous operation has completed. If the thread holding the lock performs work that requires the same lock then it will deadlock because it is trying to acquire a lock that it is already holding.
+
+If the coroutine completes on a different thread from the thread that acquired the lock then that is undefined behavior.  Even with an explicit return to the original thread an exception might be thrown before coroutine resumes and the result will be that the lock guard is not destructed.
+
+##### Example, Bad
+
+    std::mutex g_lock;
+
+    std::future<void> Class::do_something()
+    {
+        std::lock_guard<std::mutex> guard(g_lock);
+        co_await something(); // DANGER: coroutine has suspended execution while holding a lock
+        co_await somethingElse();
+    }
+
+##### Example, Good
+
+    std::mutex g_lock;
+
+    std::future<void> Class::do_something()
+    {
+        {
+            std::lock_guard<std::mutex> guard(g_lock);
+            // modify data protected by lock
+        }
+        co_await something(); // OK: lock has been released before coroutine suspends
+        co_await somethingElse();
+    }
+
+
+##### Note
+
+This pattern is also bad for performance. When a suspension point is reached, such as co_await, execution of the current function stops and other code begins to run. It may be a long period of time before the coroutine resumes. For that entire duration the lock will be held and cannot be acquired by other threads to perform work.
+
+##### Enforcement
+
+Flag all lock guards that are not destructed before a coroutine suspends.
+
+### <a name="Rcoro-reference-params"></a>CP.53: Parameters to coroutines should not be passed by reference
+
+##### Reason
+
+Once a coroutine reaches the first suspension point, such as a co_await, the synchronous portion returns. After that point any parameters passed by reference are dangling. Any usage beyond that is undefined behavior which may include writing to freed memory.
+
+##### Example, Bad
+
+    std::future<int> Class::do_something(const std::shared_ptr<int>& input)
+    {
+        co_await something();
+        
+        // DANGER: the reference to input is no longer valid and may be freed memory
```

Comment:
Shouldn't that be ".. may no longer ..."

Reply:
At this point the reference has gone bad.  The underlying object may or may not be still around but the reference is definitely past it's expiration.

Reply:
Why would the reference definetly past it's expiration? If by reference you mean the shared_ptr variable, that variable may be e.g a global variable or live beyond the lifetime of the coroutine for some other reason

Reply:
It _might_ be a global but it could also be a local variable in the calling method, in which case it is freed memory after the first suspension.  The hazard is that as the callee it doesn't know either way.  The guideline is basically saying that it is better to be safe and avoid this risk entirely.

Reply:
Which is exactly why I suggested "may" instead of "is". Maybe "probably is" 

> it could also be a local variable in the calling method, in which case it is freed memory after the first suspension.

Even that isn't necessarily the case - afaik only if the lifetime of the coroutine exceeds that of the calling function and there are several patterns where that isn't the csse (e.g. if the coroutine is a generator, or if it a coroutine itself and the two are nested).

I completely agree that it is a dangerous pattern that is most likely a bug and should be avoided, but I have a problem with the absolute claim that it "is" invalid, because itcan confuse the reader and question one's  understanding of coroutines.

Btw.: Eve

Reply:
Ah, I understand now.  That's valid.  I'll make a tweak.